### ISSUE TITLE
Read-only shared buffer (`ArrayBuffer`) shared from main process to renderer process

### ISSUE THREAD

**Issue by grovesNL** [electron/electron#45034]
### Preflight Checklist

- [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [x] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a feature request that matches the one I want to file, without success.

### Problem Description

I'm interested in repeatedly passing large amounts of data between the main and renderer processes. Currently I send binary blobs from the main process to the renderer process but this adds significant overhead. I'm interested in sharing this data more directly somehow.

### Proposed Solution

`Microsoft Edge WebView2` has introduced an API for sharing buffers in a similar way by sharing an `ArrayBuffer` and using messaging to let the renderer process know when the buffer is readable.

The documentation for the `WebView2` API is located at https://github.com/MicrosoftEdge/WebView2Feedback/blob/main/specs/SharedBuffer.md

This API seems like it would be compatible with Electron.

There is some more discussion about the API (including how it could work with atomics) in https://github.com/MicrosoftEdge/WebView2Feedback/issues/89

### Alternatives Considered

- Sending data repeatedly which is too slow for my needs
- Rendering part of the HTML in the main process, and sending the output to the renderer process instead - this is complicated because of compositing (e.g., https://github.com/electron/electron/issues/10547)

**Comment by grovesNL** [comment]
Maybe also related to #10409 which talks about a similar use case. @ymeine is `WebView2`'s `SharedBuffer` approach similar to the ideas you mentioned?

**Comment by ymeine** [comment]
@grovesNL

TL;DR: yes, WebView2’s shared buffer feature is implementing what is requested here and in issue #10409, and for which [I gave there a recipe for making your own implementations](https://github.com/electron/electron/issues/10409#issuecomment-2417414128). It is in all cases based on OS level inter process shared memory.

---

## The Electron issues

The other issue and yours actually ask for the same thing.

Issue #10409 talks about `SharedArrayBuffer` when it should talk about `ArrayBuffer` (even though the confusion is completely legitimate and I was being confused myself - see below for mmore details), but in the end is asking for inter process shared memory.

Your issue talks about `ArrayBuffer` (which is more correct) and asks for inter process shared memory as well. However, I don’t think that adding a restriction to have it read only is making it easier or safer, asking right away for a read/write access is more powerful, flexible, and technically the same: it doesn’t change the underlying technical challenge.

## Shared memory: intra process, inter process, `ArrayBuffer`, `SharedArrayBuffer` and garbage collection

### Processes, threads and memory

First of all we need to talk about the memory model in modern operating systems and hardware architectures.

They abstract memory concepts and proxy memory access (OS and CPU memory controller are involved, but I’m not an expert on that particular topic).

What that means: 

- **processes**, which correspond to program instances, are isolated from each other, and that includes memory access
- in order to still provide parallelism capabilities without the isolation issue, OSs also provide **threads**. They belong to one process and completely share the memory of that process (I won’t discuss the specific case of [thread storage duration](https://en.cppreference.com/w/cpp/language/storage_duration#Thread_storage_duration)

Now, as you already know, there are two ways to achieve Inter Process Communication (IPC): 

- **shared memory**: for efficient data sharing, but may require synchronization between processes for read/write (in the very specific use case I personally have, synchronization is not required)
- **messaging**: allows passing data through an OS layer with a synchronization between posting and delivering. As you already know, there’s an overhead, but it’s completely acceptable in many use cases, and that’s a good complementary solution for synchronizing shared memory access.

### Threads in JavaScript

While memory is naturally shared by threads within a given process, JavaScript engines do bring another layer of abstraction, and they can implement any isolation model they want since they proxy all memory-related operations.

And that’s what they do: JavaScript contexts between workers agents and main agent have isolated objects, and in the case of buffer allocation have implemented two kinds of access: 

- non-shared with `ArrayBuffer`
- shared with `SharedArrayBuffer`

But again, we are talking only about **threads** here.

### `SharedArrayBuffer` vs `ArrayBuffer` (and garbage collection)

`ArrayBuffer` and `SharedArrayBuffer`  are both regular JavaScript objects. That means that they are isolated between threads, and that they are garbage collected.

However they are just wrapper objects and their goal is to provide a particular native functionality: allocating, accessing and freeing memory blocks.

There are 3 types of behavior here: 

- **`ArrayBuffer` in pure JavaScript**: memory blocks are purposedly isolated between threads and processes. A block can be transferred from one thread to another (i.e. from one wrapper instance to another wrapper instance - the underlying memory block itself can be internally reused as is), and the implementation ensures the origin thread can’t access the memory anymore and that the target thread will be responsible for garbage collection
- **`ArrayBuffer` pointing to a custom memory address, only possible through a native addon**. This allows any kind of shared memory, and provides control over the lifetime of the pointed memory. This also provides interoperability with native libraries, it’s not always about shared memory. However, this is exactly the issue here in Electron: this is being prevented because of the *memory cage*
- **`SharedArrayBuffer`**, bringing back the natural memory sharing between threads of a process, but with a proper handling with regards to garbage collection. However, the name is misleading because it is not related to OS level inter process shared memory, it’s just “normal” shared memory between threads

## WebView2’s new feature

*Disclaimer: I actually haven’t read carefully the linked specifications document, so I may be wrong on some details. But yes: I know this is OS level shared memory.*

And so WebView2 is bringing here a 4th behavior with `ArrayBuffer` instances pointing to OS level shared memory, and with proper handling of garbage collection.

It provides APIs to explicitly deallocate the memory if needed, and to transfer the reference to the shared memory between processes, so that they have access to an `ArrayBuffer` wrapper instance pointing to that memory.

It is really like bullet point number 2 in previous section, and like what is used in the instructions I gave to create your own implementations. But having it done already is a big plus, especially when native extensions are hard or impossible to write.
